x/evmutil/keeper/msg_server_test.go

```diff
@@ -264,3 +269,314 @@ func (suite *MsgServerSuite) TestConvertERC20ToCoin() {
 		})
 	}
 }
+
+func (suite *MsgServerSuite) TestEVMCall() {
+	userAddr := app.RandomAddress()
+	userEvmAddr := common.BytesToAddress(userAddr.Bytes())
+	contractAddr := suite.DeployERC20()
+
+	authorityModule := communitytypes.ModuleAccountName
+	authorityAct := suite.AccountKeeper.GetModuleAccount(suite.Ctx, authorityModule)
+	authorityAddr := authorityAct.GetAddress().String()
+	authorityEvmAddr := types.NewInternalEVMAddress(common.BytesToAddress(authorityAct.GetAddress().Bytes()))
+	erc20StartingBal := big.NewInt(1000)
+
+	validFnAbi := `{
+		"inputs": [
+			{ "type": "address", "name": "to" },
+			{ "type": "uint256", "name": "amount" }
+		],
+		"name": "transfer",
+		"type": "function"
+	}`
+	validData := fmt.Sprintf(
+		"0x%s%s%s",
+		"a9059cbb", // transfer(address,uint256)
+		hexutil.Encode(common.LeftPadBytes(userEvmAddr.Bytes(), 32))[2:],
+		hexutil.Encode(common.LeftPadBytes(big.NewInt(30).Bytes(), 32))[2:],
+	)
+
+	type errArgs struct {
+		expectPass bool
+		contains   string
+	}
+
+	tests := []struct {
+		name       string
+		msg        types.MsgEVMCall
+		expUsdcBal sdk.Int
+		errArgs    errArgs
+	}{
+		{
+			"valid - erc20 contract transfer call",
+			types.MsgEVMCall{
+				To:        contractAddr.String(),
+				FnAbi:     validFnAbi,
+				Data:      validData,
+				Authority: authorityAddr,
+				Amount:    sdk.ZeroInt(),
+			},
+			sdk.NewInt(970),
+			errArgs{
+				expectPass: true,
+			},
+		},
+		{
+			"valid - transfer call",
+			types.MsgEVMCall{
+				To:        userEvmAddr.String(),
+				Amount:    keeper.ConversionMultiplier.Mul(sdk.NewInt(11)),
+				Authority: authorityAddr,
+			},
+			sdk.NewInt(1000),
+			errArgs{
+				expectPass: true,
+			},
+		},
+		{
+			"valid - payable deposit",
+			types.MsgEVMCall{
+				To:     userEvmAddr.String(),
+				Amount: keeper.ConversionMultiplier.Mul(sdk.NewInt(11)),
+				FnAbi: `{
+					"inputs": [],
+					"name": "deposit",
+					"type": "function"
+				}`,
+				Data: fmt.Sprintf(
+					"0x%s",
+					"d0e30db0", // deposit()
+				),
+				Authority: authorityAddr,
+			},
+			sdk.NewInt(1000),
+			errArgs{
+				expectPass: true,
+			},
+		},
+		{
+			"valid - extra data passed after valid fn data",
```

Comment:
Yeah this was a surprise to me as I originally had this as a failure test. But digging through, this is all how Geth handles contract calls. It allows the extra data to be passed but the contract function will just ignore it (similar to JS). The extra args must also need to be properly decoded, otherwise hex decoding will fail. It is also not possible to use this to cause issues with existing params since they are all fixed length. Would be great to also brainstorm additional issues that can be caused from this.

Reply:
https://github.com/Kava-Labs/kava/pull/1405/commits/6813d79bc43a089872782c9e40147e17c75cfe05
Added some changes to verify the arg numbers in `Decode` & updated our tests. Now passing in extra arguments should fail.